File: backend/app/schemas/notifications.py
"""Notification API schemas."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class _Schema(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class NotificationSchema(_Schema):
    id: uuid.UUID
    category: str
    severity: str
    title: str
    message: str
    is_read: bool
    read_at: datetime | None
    payload: dict | None
    created_at: datetime

    @classmethod
    def from_model(cls, model) -> NotificationSchema:  # noqa: ANN001
        return cls(
            id=model.id,
            category=model.type,
            severity=model.severity.value
            if hasattr(model.severity, "value")
            else str(model.severity),
            title=model.title,
            message=model.message,
            is_read=model.is_read,
            read_at=model.read_at,
            payload=model.payload,
            created_at=model.created_at,
        )


class NotificationPageSchema(_Schema):
    items: list[NotificationSchema]
    total: int
    page: int
    page_size: int


class UnreadCountSchema(_Schema):
    unread: int

File: web/app/services/api/adapters/notifications.ts
import type { Notification, NotificationPage, NotificationSeverity } from '~/types/notifications'
import { resourceLinkFromPayload } from '~/utils/resourceLinks'
import { iso, isoOrNull, normalizeEnvironment } from './common'
import type { RawNotification, RawNotificationPage } from './raw'

export function toNotification(raw: RawNotification): Notification {
  return {
    id: raw.id,
    category: raw.category,
    severity: raw.severity as NotificationSeverity,
    title: raw.title,
    message: raw.message,
    is_read: raw.is_read,
    read_at: isoOrNull(raw.read_at),
    payload: raw.payload,
    created_at: iso(raw.created_at),
    link: resourceLinkFromPayload(raw.payload),
    environment: normalizeEnvironment(raw.payload),
  }
}

export function toNotificationPage(raw: RawNotificationPage): NotificationPage {
  return {
    items: raw.items.map(toNotification),
    total: raw.total,
    page: raw.page,
    pageSize: raw.page_size,
  }
}
